### DISCUSSION TITLE
(pageRendered) firing before page is complete rendered?

### DISCUSSION THREAD

**Discussion by othon-lima** [stephanrauh/ngx-extended-pdf-viewer#1548] [Q&A]
Hello!

I'm hoping to get some insight in a problem that I can't seem to overcome.

I want to get the coordinates for every highlighted word - in every page if "Highlight All" is selected, using the built-in search tool. The approach I took was to first loop through the document and render all the pages, after that I can loop again through the pages - already rendered at this point, and get the highlighted elements by their IDs. I am aware this is not the most optimal way, but I'm just trying to get something working for now and hopefully later on refactor to something more elegant. Feel free to suggest a different approach. I'm quite new, so trying to work with what I have.

I am very close to achieving that.  However, it seems like the (pageRendered) is firing before the page finished rendering. I say that because I have a list of rendered pages, that gets updated every time the method called by the (pageRendered) event is triggered. It is important to remember that the pdf viewer is also rendering the next page. After this list length is equal to the number of pages found in the document then I loop though the rendered pages getting the highlighted elements. It is currently working for every page, except the last rendered page. I don't know what other event I could try to use to ensure all pages are indeed rendered.

Any help is very much welcomed and appreciated!

**Comment by stephanrauh** [comment]
Try `(textLayerRendered)`. The other event is fired when the image has been rendered.

**Reply by othon-lima** [reply]
Hi Stephan,

As always, thank you for the quick reply.

I actually tried (textLayerRendered) right after making my original post and before reading your suggestion...still no luck, same issue.  
It is weird, if I put a small timer right after the last rendered page is detected, then it works. 

Is there any way I can clear the viewer renderer cache and only render a page at a time? This way I can ensure that, when needed, every time I iterate through a page, I'll get a pageRendered back to that page only when said page is opened.

**Reply by stephanrauh** [reply]
Adding a small timeout helps surprisingly often. :) 
How to clear the renderer cache is a good question. I've quickly browsed the source code. Currently, there seems to be no such setting. Until now, I was only interesting in doing the opposite thing (see https://pdfviewer.net/extended-pdf-viewer/prerendering). On the other hand, pdf.js does remove pages after a while, so it should be possible to find and tweak the algorithm. It's probably in the base_viewer.js file.

**Reply by othon-lima** [reply]
Sounds good, I'll take a look at that.

Last question, I promise. After a page is rendered, if I change pages and then go back to the already rendered page, there is no event emitted once the change page action is completed, right?

**Reply by stephanrauh** [reply]
Sure there is. You're looking for the `(page)` event. :)
